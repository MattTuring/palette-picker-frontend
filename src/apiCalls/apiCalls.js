export const getProjects = async () => {
  const url = 'https://palette-pick-backend.herokuapp.com/api/v1/projects';
  const response = await fetch(url);
  if (!response.ok) {
    throw new Error('There was an error getting projects.');
  }
  const projects = await response.json();
  return projects
}

export const getPalettes = async (id) => {
  const url = `https://palette-pick-backend.herokuapp.com/api/v1/projects/${id}/palettes`
  const response = await fetch(url);
  if (!response.ok) {
    throw new Error('There was an error getting palettes.');
  }
  const palettes = await response.json();
  return palettes
}

export const postProject = async (project) => {
  const url = 'https://palette-pick-backend.herokuapp.com/api/v1/projects';
  const options = {
    method: 'POST',
    body: JSON.stringify(project),
    headers: {
      'Content-Type': 'application/json'
    }
  }
  const response = await fetch(url, options);

  if (!response.ok) {
    throw new Error('There was an error posting project.');
  }

  const result = await response.json();
  return result;
}

export const postPalette = async (palette) => {
  const url = 'https://palette-pick-backend.herokuapp.com/api/v1/palettes';
  const options = {
    method: 'POST',
    body: JSON.stringify(palette),
    headers: {
      'Content-Type': 'application/json'
    }
  }
  const response = await fetch(url, options);

  if (!response.ok) {
    throw new Error('There was an error posting palette.');
  }

  const result = await response.json();
  return result;
}

export const updateProject = async (project, id) => {
  const url = `https://palette-pick-backend.herokuapp.com/api/v1/projects/${id}`;
  const options = {
    method: 'PUT',
    body: JSON.stringify(project),
    headers: {
      'Content-Type': 'application/json'
    }
  }
  const response = await fetch(url, options);

  if (!response.ok) {
    throw new Error('There was an error updating project.');
  }

  const result = await response.json();
  return result;
}

export const updatePalette = async (palette, id) => {
  const url = `https://palette-pick-backend.herokuapp.com/api/v1/palettes/${id}`;
  const options = {
    method: 'PUT',
    body: JSON.stringify(palette),
    headers: {
      'Content-Type': 'application/json'
    }
  }
  const response = await fetch(url, options);

  if (!response.ok) {
    throw new Error('There was an error updating palette.');
  }

  const result = await response.json();
  return result;
}

export const deleteProject = async (id) => {
  const url = `https://palette-pick-backend.herokuapp.com/api/v1/projects/${id}`;
  const options = {
    method: 'DELETE',
    headers: {
      'Content-Type': 'application/json'
    }
  }
  const response = await fetch(url, options);

  if (!response.ok) {
    throw new Error('There was an error deleting project.');
  }

  const result = await response.json();
  return result;
}

export const deletePalette = async (id) => {
  const url = `https://palette-pick-backend.herokuapp.com/api/v1/palettes/${id}`;
  const options = {
    method: 'DELETE',
    headers: {
      'Content-Type': 'application/json'
    }
  }
  const response = await fetch(url, options);

  if (!response.ok) {
    throw new Error('There was an error deleting palette.');
  }

  const result = await response.json();
  return result;
}
