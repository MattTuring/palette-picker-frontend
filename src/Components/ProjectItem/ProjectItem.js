import React, { useState, useEffect } from 'react';
import { deleteProject, getPalettes, updateProject } from '../../apiCalls/apiCalls';
import trash from '../../trash.png';
import edit from '../../edit.png';
import ProjectPaletteItem from '../ProjectPaletteItem/ProjectPaletteItem'

const ProjectItem = ({project, projects, setProjects}) => {
  let [palettes, setPalettes] = useState([]);
  const [name, setName] = useState('');
  useEffect(() => setName(project.name), []);

  const removeProject = async () => {
    try {
      const newProjects = projects.filter(prj => prj.id !== project.id);
      setProjects(newProjects);
      await deleteProject(project.id);
    }
    catch (error) {
      console.log(error);
    }
  };

  const getPalettesForProject = async () => {
    try {
      const receivedPalettes = await getPalettes(project.id);
      setPalettes(receivedPalettes);
    }
    catch (error) {
      console.log(error);
    }
  };

  const paletteItems = palettes.map((palette, ind) => (
    <ProjectPaletteItem
    key={Date.now() + palette.name + ind}
    {...{palettes, setPalettes, palette}} />
  ));


  let [newProjectName, setProjectNewName] = useState('');
  let [showProjectNewName, setProjectShowNewName] = useState(false);

  return (
    <>
      <h4 key={project.id}>{name}
        <img
          src={trash}
          onClick={removeProject}
          className="delete"
          alt='delete'/>
          <img src={edit} onClick={() => setProjectShowNewName(!showProjectNewName)} className="edit" alt='edit'/>
          {showProjectNewName && <><input onChange={(event) => setProjectNewName(event.target.value)} type='text' placeholder='New Project Name'/><button onClick={() => { updateProject({name: newProjectName},project.id); setName(newProjectName)}}>Update Name</button></>}
      </h4>
        <button onClick={getPalettesForProject}>
          Show/Update Palettes
        </button>
      {paletteItems}
    </>
  )
};

export default ProjectItem;
