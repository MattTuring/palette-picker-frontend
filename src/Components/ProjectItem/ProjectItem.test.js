import React from 'react';
import { shallow } from 'enzyme';
import ProjectItem from './ProjectItem';
import { deleteProject, getPalettes } from '../../apiCalls/apiCalls';

jest.mock('../../apiCalls/apiCalls');

describe('ProjectItem', () => {
  const setProjects = jest.fn();

  const mockProps = {
    project: {
      id: 1,
      name: 'Project'
    },
    projects: [{
      id: 1,
      name: 'Project'
    }],
    setProjects
  };

  let wrapper;

  beforeEach(() => {
    wrapper = shallow(<ProjectItem {...mockProps} />);
  });

  it('should match the snapshot with all data passed in correctly', () => {
    expect(wrapper).toMatchSnapshot();
  });

  it('should call setProjects after click on remove button', () => {
    wrapper.find('.delete').simulate('click');
    expect(setProjects).toHaveBeenCalledWith([]);
  });

  it('should call deleteProject after click on remove button', () => {
    wrapper.find('.delete').simulate('click');
    expect(deleteProject).toHaveBeenCalledWith(1);
  });

  it('should call getPalettes after click on show button', () => {
    getPalettes.mockImplementation(() => {
      return Promise.resolve([]);
    });

    wrapper.find('button').simulate('click');
    expect(getPalettes).toHaveBeenCalledWith(1);
  });
});
