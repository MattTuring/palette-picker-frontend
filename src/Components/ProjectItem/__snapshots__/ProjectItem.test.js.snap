// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ProjectItem should match the snapshot with all data passed in correctly 1`] = `
<Fragment>
  <h4
    key="1"
  >
    Project
    <img
      alt="delete"
      className="delete"
      onClick={[Function]}
      src="trash.png"
    />
    <img
      alt="edit"
      className="edit"
      src="edit.png"
    />
  </h4>
  <button
    onClick={[Function]}
  >
    Show/Update Palettes
  </button>
</Fragment>
`;
